# 01. GTSRB Project Setup and Data Check

In this notebook, we will start the **German Traffic Sign Recognition** project using PyTorch.

The goal of this notebook is not to train a model yet.  
The goal is to make sure our environment, project folders, and dataset files are ready before we begin EDA and modeling.

## What we're going to cover

1. Check Python, PyTorch, and CUDA setup
2. Check project folder paths
3. Check whether dataset files exist
4. Load the CSV files
5. Inspect the basic structure of the dataset
6. Decide what to explore next

> **Rule:** We will move step by step. We will not start modeling before understanding the setup and data.

## 0. Getting setup

Before working with the dataset, we need to check whether our Python environment and PyTorch installation are working correctly.

### Guiding questions

- Which Python version are we using?
- Which PyTorch version are we using?
- Is CUDA/GPU available?
- If CUDA is available, which GPU is being used?

In [1]:
import sys
import torch
import torchvision

print(f"Python version: {sys.version}")
print(f"PyTorch version: {torch.__version__}")
print(f"Torchvision version: {torchvision.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")

if torch.cuda.is_available():
    print(f"CUDA device name: {torch.cuda.get_device_name(0)}")
    print(f"CUDA version used by PyTorch: {torch.version.cuda}")

Python version: 3.11.15 | packaged by conda-forge | (main, Jun 11 2026, 03:27:10) [MSC v.1944 64 bit (AMD64)]
PyTorch version: 2.11.0+cu128
Torchvision version: 0.26.0+cu128
CUDA available: True
CUDA device name: NVIDIA GeForce RTX 5060 Ti
CUDA version used by PyTorch: 12.8


### Observation

Our Python and PyTorch environment is working correctly.

- We are using **Python 3.11.15**
- We are using **PyTorch 2.11.0**
- We are using **Torchvision 0.26.0**
- **CUDA is available**
- PyTorch is able to detect the GPU successfully
- The current GPU is **NVIDIA GeForce RTX 5060 Ti**
- PyTorch is using **CUDA 12.8**

This means the environment setup is complete and ready for deep learning experiments.

### Decision

Since the software and GPU setup are working correctly, the next step is to check the **project folder structure** and confirm that the **dataset files and folders** are in the expected location.

## 1. Check project paths and folder structure

Before loading the dataset, we need to make sure the notebook is running from the correct location and that the main project folders exist.

### Guiding questions

- What is the current working directory?
- What is the project root directory?
- Does the `Data/` folder exist?
- Which main folders and files are currently available in the project root?

In [2]:
from pathlib import Path
import os

# Current notebook location
current_dir = Path.cwd()

# If notebook is inside "Notebook", then project root is its parent
if current_dir.name.lower() == "notebook":
    project_root = current_dir.parent
else:
    project_root = current_dir

data_dir = project_root / "Data"

print("Current working directory:", current_dir)
print("Project root directory:", project_root)
print("Data directory path:", data_dir)
print("Does Data directory exist?:", data_dir.exists())

print("\nTop-level items in project root:")
for item in sorted(project_root.iterdir()):
    print("-", item.name)

Current working directory: c:\Users\USER\Kawsar_Ahmmed\ALL_Projects_Lab\Deep_Learning\Pytorch-German-Traffic-Sign-Recognition\Notebook
Project root directory: c:\Users\USER\Kawsar_Ahmmed\ALL_Projects_Lab\Deep_Learning\Pytorch-German-Traffic-Sign-Recognition
Data directory path: c:\Users\USER\Kawsar_Ahmmed\ALL_Projects_Lab\Deep_Learning\Pytorch-German-Traffic-Sign-Recognition\Data
Does Data directory exist?: True

Top-level items in project root:
- .git
- .gitignore
- Data
- LICENSE
- Notebook
- Processed_Data
- README.md
- Report
- Trained_model


## 2. Check dataset files inside the Data folder

Now we want to inspect the `Data/` folder more closely.

### Guiding questions

- Which files and folders are available inside `Data/`?
- Are there CSV files such as `Train.csv`, `Test.csv`, or `Meta.csv`?
- Are there image folders such as `Train/`, `Test/`, or `Meta/`?
- Is the dataset layout ready for EDA?

In [3]:
if data_dir.exists():
    print("Items inside Data folder:\n")
    for item in sorted(data_dir.iterdir()):
        item_type = "Folder" if item.is_dir() else "File"
        print(f"- {item.name} ({item_type})")
else:
    print("Data folder was not found.")

Items inside Data folder:

- Meta (Folder)
- Meta.csv (File)
- Test (Folder)
- Test.csv (File)
- Train (Folder)
- Train.csv (File)


In [4]:
if data_dir.exists():
    csv_files = sorted(data_dir.glob("*.csv"))
    subfolders = sorted([item for item in data_dir.iterdir() if item.is_dir()])

    print("CSV files found:")
    if csv_files:
        for file in csv_files:
            print("-", file.name)
    else:
        print("No CSV files found in Data/")

    print("\nSubfolders found:")
    if subfolders:
        for folder in subfolders:
            print("-", folder.name)
    else:
        print("No subfolders found in Data/")
else:
    print("Data folder was not found.")

CSV files found:
- Meta.csv
- Test.csv
- Train.csv

Subfolders found:
- Meta
- Test
- Train


### Observation

The notebook is running from the `Notebook/` folder, and the project root was detected correctly.

The main project folders are available:

- `Data/`
- `Notebook/`
- `Processed_Data/`
- `Report/`
- `Trained_model/`

Inside the `Data/` folder, the expected GTSRB dataset files and folders are available:

- `Train/`
- `Test/`
- `Meta/`
- `Train.csv`
- `Test.csv`
- `Meta.csv`

This means the dataset is available in the correct location and the project folder structure is ready.

### Decision

Since the dataset files are available, the next step is to load the CSV files and inspect their basic structure before working with the images.

## 3. Load dataset CSV files

The GTSRB dataset includes CSV files that contain important information about the images and labels.

Before loading images directly, we will inspect these CSV files to understand how the dataset is organized.

### Guiding questions

- What columns are available in `Train.csv`, `Test.csv`, and `Meta.csv`?
- How many rows are in each CSV file?
- Which column contains the image path?
- Which column contains the class label?

In [5]:
import pandas as pd

train_csv_path = data_dir / "Train.csv"
test_csv_path = data_dir / "Test.csv"
meta_csv_path = data_dir / "Meta.csv"

train_df = pd.read_csv(train_csv_path)
test_df = pd.read_csv(test_csv_path)
meta_df = pd.read_csv(meta_csv_path)

print("Train CSV shape:", train_df.shape)
print("Test CSV shape:", test_df.shape)
print("Meta CSV shape:", meta_df.shape)

Train CSV shape: (39209, 8)
Test CSV shape: (12630, 8)
Meta CSV shape: (43, 5)


In [6]:
print("Train CSV columns:")
print(train_df.columns.tolist())

print("\nTest CSV columns:")
print(test_df.columns.tolist())

print("\nMeta CSV columns:")
print(meta_df.columns.tolist())

Train CSV columns:
['Width', 'Height', 'Roi.X1', 'Roi.Y1', 'Roi.X2', 'Roi.Y2', 'ClassId', 'Path']

Test CSV columns:
['Width', 'Height', 'Roi.X1', 'Roi.Y1', 'Roi.X2', 'Roi.Y2', 'ClassId', 'Path']

Meta CSV columns:
['Path', 'ClassId', 'ShapeId', 'ColorId', 'SignId']


In [7]:
train_df.head()

,Width,Height,Roi.X1,Roi.Y1,Roi.X2,Roi.Y2,ClassId,Path
0,27,26,5,5,22,20,20,Train/20/00020_00000_00000.png
1,28,27,5,6,23,22,20,Train/20/00020_00000_00001.png
2,29,26,6,5,24,21,20,Train/20/00020_00000_00002.png
3,28,27,5,6,23,22,20,Train/20/00020_00000_00003.png
4,28,26,5,5,23,21,20,Train/20/00020_00000_00004.png


In [8]:
test_df.head()

,Width,Height,Roi.X1,Roi.Y1,Roi.X2,Roi.Y2,ClassId,Path
0,53,54,6,5,48,49,16,Test/00000.png
1,42,45,5,5,36,40,1,Test/00001.png
2,48,52,6,6,43,47,38,Test/00002.png
3,27,29,5,5,22,24,33,Test/00003.png
4,60,57,5,5,55,52,11,Test/00004.png


In [9]:
meta_df.head()

,Path,ClassId,ShapeId,ColorId,SignId
0,Meta/27.png,27,0,0,1.32
1,Meta/0.png,0,1,0,3.29
2,Meta/1.png,1,1,0,3.29
3,Meta/10.png,10,1,0,3.27
4,Meta/11.png,11,0,0,1.22


### Observation

The dataset CSV files were loaded successfully.

- `Train.csv` has **39,209 rows** and **8 columns**
- `Test.csv` has **12,630 rows** and **8 columns**
- `Meta.csv` has **43 rows** and **5 columns**

Both `Train.csv` and `Test.csv` contain the same important columns:

- `Width`
- `Height`
- `Roi.X1`, `Roi.Y1`, `Roi.X2`, `Roi.Y2`
- `ClassId`
- `Path`

The most important columns for this project are:

- `Path`: image file path
- `ClassId`: target label/class

`Meta.csv` contains extra information about the 43 traffic sign classes.

### Decision

The CSV structure looks correct.  
Before ending this setup notebook, we will do one basic integrity check for missing values, duplicate rows, class count, and sample image paths.

## 4. Basic dataset integrity check

Before finishing this notebook, we will run a small sanity check on the CSV files.

### Guiding questions

- Are there any missing values?
- Are there duplicate rows?
- How many unique classes are available?
- Do the image paths look valid?

In [10]:
print("Missing values in Train.csv:")
print(train_df.isna().sum())

print("\nMissing values in Test.csv:")
print(test_df.isna().sum())

print("\nDuplicate rows:")
print("Train duplicates:", train_df.duplicated().sum())
print("Test duplicates:", test_df.duplicated().sum())

print("\nUnique classes:")
print("Train unique classes:", train_df["ClassId"].nunique())
print("Test unique classes:", test_df["ClassId"].nunique())
print("Meta unique classes:", meta_df["ClassId"].nunique())

print("\nSample train image paths:")
print(train_df["Path"].head().tolist())

print("\nSample test image paths:")
print(test_df["Path"].head().tolist())

Missing values in Train.csv:
Width      0
Height     0
Roi.X1     0
Roi.Y1     0
Roi.X2     0
Roi.Y2     0
ClassId    0
Path       0
dtype: int64

Missing values in Test.csv:
Width      0
Height     0
Roi.X1     0
Roi.Y1     0
Roi.X2     0
Roi.Y2     0
ClassId    0
Path       0
dtype: int64

Duplicate rows:
Train duplicates: 0
Test duplicates: 0

Unique classes:
Train unique classes: 43
Test unique classes: 43
Meta unique classes: 43

Sample train image paths:
['Train/20/00020_00000_00000.png', 'Train/20/00020_00000_00001.png', 'Train/20/00020_00000_00002.png', 'Train/20/00020_00000_00003.png', 'Train/20/00020_00000_00004.png']

Sample test image paths:
['Test/00000.png', 'Test/00001.png', 'Test/00002.png', 'Test/00003.png', 'Test/00004.png']


### Observation

The basic dataset integrity check looks clean.

- `Train.csv` has **no missing values**
- `Test.csv` has **no missing values**
- There are **no duplicate rows** in both train and test CSV files
- Train, test, and meta files all contain **43 unique classes**
- Image paths are available in the `Path` column
- Train image paths follow this structure: `Train/class_id/image_name.png`
- Test image paths follow this structure: `Test/image_name.png`

This means the CSV files are clean enough to continue with image-level exploration.

### Decision

The setup and basic dataset checks are complete.  
In the next notebook, we will start proper EDA by exploring:

- class distribution
- image sizes
- random image samples
- one image per class
- possible class imbalance
- visual quality of traffic sign images

We will still not train a model until the data is properly understood.